### PR TITLE
[SPARK-41230][CONNECT][PYTHON] Remove `str` from Aggregate expression type

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

This PR proposes that Relations (e.g. Aggregate in this PR) should only deal with  `Expression` than `str`. `str` could be mapped to different expressions (e.g. sql expression, unresolved_attribute, etc.). Relations are not supposed to understand the difference of `str` but DataFrame should understand it.

This PR specifically changes for `Aggregate`.

### Why are the changes needed?

Codebase refactoring.

### Does this PR introduce _any_ user-facing change?

 No

### How was this patch tested?

UT

### CODE REVIEW

**Comment by zhengruifeng** [comment]
I think it is time to refactor `Column` to something like:

```
class Column:

    def __init__(self, expr: Expression) -> None:
        self.expr = expr
    ...
```

we'd better try to match the API shape at first, otherwise it will be hard to do so after we make it more complicated.

cc @HyukjinKwon

File: python/pyspark/sql/connect/dataframe.py [hunk 1]
```diff
@@ -79,23 +67,23 @@ def agg(self, exprs: Optional[MeasuresType] = None) -> "DataFrame":
         )
         return res
 
-    def _map_cols_to_dict(self, fun: str, cols: List[Union[Column, str]]) -> Dict[str, str]:
-        return {x if isinstance(x, str) else x.name(): fun for x in cols}
+    def _map_cols_to_expression(self, fun: str, col: Union[Column, str]) -> Sequence[Expression]:
```
Comment (grundprinzip) [inline python/pyspark/sql/connect/dataframe.py:70]:
```suggestion
    def _map_cols_to_expression(self, fun: str, col: Union[Expression, str]) -> Sequence[Expression]:
```

File: python/pyspark/sql/connect/dataframe.py [hunk 2]
```diff
@@ -79,23 +67,23 @@ def agg(self, exprs: Optional[MeasuresType] = None) -> "DataFrame":
...
+        return [ScalarFunctionExpression(fun, Column(col) if isinstance(col, str) else col)]
 
-    def min(self, *cols: Union[Column, str]) -> "DataFrame":
-        expr = self._map_cols_to_dict("min", list(cols))
+    def min(self, col: Union[Column, str]) -> "DataFrame":
```
Comment (grundprinzip) [inline python/pyspark/sql/connect/dataframe.py:73]:
```suggestion
    def min(self, col: Union[Expression, str]) -> "DataFrame":
```

File: python/pyspark/sql/connect/dataframe.py [hunk 3]
```diff
@@ -79,23 +67,23 @@ def agg(self, exprs: Optional[MeasuresType] = None) -> "DataFrame":
...
+        expr = self._map_cols_to_expression("min", col)
         return self.agg(expr)
 
-    def max(self, *cols: Union[Column, str]) -> "DataFrame":
-        expr = self._map_cols_to_dict("max", list(cols))
+    def max(self, col: Union[Column, str]) -> "DataFrame":
```
Comment (grundprinzip) [inline python/pyspark/sql/connect/dataframe.py:77]:
```suggestion
    def max(self, col: Union[Expression, str]) -> "DataFrame":
```

File: python/pyspark/sql/connect/dataframe.py [hunk 4]
```diff
@@ -79,23 +67,23 @@ def agg(self, exprs: Optional[MeasuresType] = None) -> "DataFrame":
...
+        expr = self._map_cols_to_expression("max", col)
         return self.agg(expr)
 
-    def sum(self, *cols: Union[Column, str]) -> "DataFrame":
-        expr = self._map_cols_to_dict("sum", list(cols))
+    def sum(self, col: Union[Column, str]) -> "DataFrame":
```
Comment (grundprinzip) [inline python/pyspark/sql/connect/dataframe.py:81]:
```suggestion
    def sum(self, col: Union[Expression, str]) -> "DataFrame":
```

File: python/pyspark/sql/connect/dataframe.py [hunk 5]
```diff
@@ -164,8 +152,21 @@ def selectExpr(self, *expr: Union[str, List[str]]) -> "DataFrame":
 
         return DataFrame.withPlan(plan.Project(self._plan, *sql_expr), session=self._session)
 
-    def agg(self, exprs: Optional[GroupingFrame.MeasuresType]) -> "DataFrame":
-        return self.groupBy().agg(exprs)
+    def agg(self, *exprs: Union[Expression, Dict[str, str]]) -> "DataFrame":
+        if not exprs:
+            raise ValueError("exprs should not be empty")
```
Comment (grundprinzip) [inline python/pyspark/sql/connect/dataframe.py:157]:
```suggestion
            raise ValueError("Argument 'exprs' must not be empty")
```

File: python/pyspark/sql/connect/plan.py [hunk 6]
```diff
@@ -558,29 +557,19 @@ def _repr_html_(self) -> str:
...
 class Aggregate(LogicalPlan):
-    MeasureType = Tuple["ExpressionOrString", str]
-    MeasuresType = Sequence[MeasureType]
-    OptMeasuresType = Optional[MeasuresType]
-
     def __init__(
         self,
         child: Optional["LogicalPlan"],
         grouping_cols: List[Column],
-        measures: OptMeasuresType,
```
Comment (grundprinzip) [inline python/pyspark/sql/connect/plan.py:569]: There is no longer a way to call this with empty measures?

**Review by grundprinzip — COMMENTED** [review]
I think directional you're right this is the better approach. Plan object should only deal with other plans and expressions, conversion to expressions should happen before the plan is constructed.

**Comment by amaliujia** [comment]
@zhengruifeng sure maybe we can do that refactoring in this PR directly.

the plain `str` passing through to relations became a blocker for that refactoring (given there is no place for a plain `str` in expression system, it must be wrapped).

File: python/pyspark/sql/connect/plan.py (hunk 6 above)
Comment (amaliujia) [reply python/pyspark/sql/connect/plan.py:569]: This is a sequence now and I think it can be len=0 which is empty measures?

**Comment by amaliujia** [comment]
@zhengruifeng @grundprinzip can you take another look?

File: python/pyspark/sql/connect/dataframe.py [hunk 7]
```diff
@@ -164,8 +156,20 @@ def selectExpr(self, *expr: Union[str, List[str]]) -> "DataFrame":
...
-    def agg(self, exprs: Optional[GroupingFrame.MeasuresType]) -> "DataFrame":
-        return self.groupBy().agg(exprs)
+    def agg(self, *exprs: Union[Expression, Dict[str, str]]) -> "DataFrame":
+        if not exprs:
+            raise ValueError("Argument 'exprs' must not be empty")
+
+        if len(exprs) == 1 and isinstance(exprs[0], dict):
+            measures = []
+            for e, fun in exprs[0].items():
+                measures.append(ScalarFunctionExpression(fun, Column(e)))
```
Comment (zhengruifeng) [inline python/pyspark/sql/connect/dataframe.py:166]:
```suggestion
            measures = [ScalarFunctionExpression(f, Column(e)) for e, f in exprs[0].items()]
```

File: python/pyspark/sql/connect/dataframe.py [hunk 8]
```diff
@@ -51,24 +52,11 @@
 
 
 class GroupingFrame(object):
```
Comment (zhengruifeng) [inline python/pyspark/sql/connect/dataframe.py:54]: not related to this PR, but shall we rename it `GroupedData` to be the same with pyspark?

**Review by grundprinzip — APPROVED** [review]
One nit.

**Comment by HyukjinKwon** [comment]
Re: https://github.com/apache/spark/pull/38768?notification_referrer_id=NT_kwDOAGLXhbI0OTA0MTUyNTY4OjY0Nzc3MDE&notifications_query=is%3Aunread#issuecomment-1324694898

Yes, please let's refactor as we offline discussed. This refactoring has to be done before starting working on functions cc @xinrong-meng too.

**Review by HyukjinKwon — APPROVED** [review]

File: python/pyspark/sql/connect/dataframe.py (hunk 8 above)
Comment (amaliujia) [reply python/pyspark/sql/connect/dataframe.py:54]: Renamed

File: python/pyspark/sql/connect/dataframe.py [hunk 9]
```diff
@@ -51,24 +52,11 @@
...
-        # Normalize the dictionary into a list of tuples.
-        if isinstance(exprs, Dict):
-            measures = list(exprs.items())
-        elif isinstance(exprs, List):
-            measures = exprs
-        else:
-            measures = []
-
+    def agg(self, measures: Sequence[Expression]) -> "DataFrame":
         res = DataFrame.withPlan(
```
[the remark this answers is not in the document]
Comment (amaliujia) [reply python/pyspark/sql/connect/dataframe.py:60]: added.

**Comment by amaliujia** [comment]
This PR is a blocker for the refactoring. I will refactor based on the change in this PR, which will further unblocks functions.

**Comment by zhengruifeng** [comment]
merged into master

**Pull request closed without merging**